File: client/GhostClient/Sprites/NetworkPlayer.cs
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using Ghost.Sprites.Effects;
using Microsoft.Xna.Framework.Graphics;
using Sharp2D;

namespace Ghost
{
    public class NetworkPlayer : Entity
    {
        private const byte MaxLives = 3;

        private Sprite[] lifeBalls;
        private byte lives = 3;
        private bool dead = false;
        public string Username { get; private set; }

        public byte Lives
        {
            get { return lives; }
            set
            {
                lives = value;
                UpdateLifeBalls();
            }
        }

        public override float Alpha
        {
            get { return base.Alpha; }
            set
            {
                base.Alpha = value;

                if (lifeBalls == null) return;
                foreach (Sprite s in lifeBalls.Where(s => s != null))
                {
                    s.Alpha = value;
                }
            }
        }

        private Color oColor;
        public bool IsDead
        {
            get { return dead; }
            set
            {
                if (!dead && value)
                {
                    oColor = TintColor;
                    TintColor = System.Drawing.Color.FromArgb(234, 234, 32);
                    Alpha = 0.6f;
                }
                else if (dead && !value)
                {
                    TintColor = oColor;
                }
                dead = value;
            }
        }

        public List<OrbitEffect> Orbits = new List<OrbitEffect>();

        public NetworkPlayer(short id, string name) : base(id)
        {
            Username = name;
        }

        protected override void OnLoad()
        {
            Texture = GhostClient.Ghost.CurrentGhostGame.Content.Load<Texture2D>("sprites/ball.png");
            DepthTexture = GhostClient.Ghost.CurrentGhostGame.Content.Load<Texture2D>("sprites/ball.png");
            NormalTexture = GhostClient.Ghost.CurrentGhostGame.Content.Load<Texture2D>("sprites/ball_normal.jpg");

            Width = Texture.Width;
            Height = Texture.Height;

            UniformScale = 0.75f;

            NeverClip = true;
        }

        protected override void OnUnload()
        {
        }

        protected override void OnDispose()
        {
        }

        protected override void OnDisplay()
        {
            UpdateLifeBalls();
        }

        protected virtual void UpdateLifeBalls()
        {
            if (lifeBalls != null)
            {
                foreach (Sprite t in lifeBalls.Where(t => t != null))
                {
                    Deattach(t);
                    CurrentWorld.RemoveSprite(t);
                }
                lifeBalls = null;
            }

            lifeBalls = new Sprite[MaxLives];
            for (int i = 0; i < lives; i++)
            {
                Sprite temp = Sprite.FromImage("sprites/ball.png");
                temp.UniformScale = 0.2F;
                temp.NeverClip = true;
                temp.X = X - ((Width / 1.5f) / 2f);
                temp.Y = Y + 40f;
                temp.TintColor = System.Drawing.Color.FromArgb(255, 20, 183, 52);
                temp.Alpha = Alpha;

                temp.X += (((Width / 1.5f) / (MaxLives - 1)) * i);
                lifeBalls[i] = temp;

                CurrentWorld.AddSprite(lifeBalls[i]);

                Attach(temp);
            }
        }
    }
}

File: client/core/src/me/eddiep/ghost/client/core/game/sprites/NetworkPlayer.kt
package me.eddiep.ghost.client.core.game.sprites

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.math.Vector3
import com.badlogic.gdx.physics.box2d.Body
import com.badlogic.gdx.physics.box2d.BodyDef
import com.badlogic.gdx.physics.box2d.CircleShape
import me.eddiep.ghost.client.Ghost
import me.eddiep.ghost.client.core.animations.AnimationType
import me.eddiep.ghost.client.core.game.Direction
import me.eddiep.ghost.client.core.game.Entity
import me.eddiep.ghost.client.core.game.sprites.effects.OrbitEffect
import me.eddiep.ghost.client.utils.Constants
import java.util.*
import kotlin.properties.Delegates

open class NetworkPlayer(id: Short, name: String) : Entity("sprites/ball.png", id) {
    val orbits: ArrayList<OrbitEffect> = ArrayList()
    var frozen: Boolean = false
    var lastDirection = Direction.DOWN
    lateinit var body: Body;

    var lives : Byte by Delegates.observable(3.toByte()) {
        d, old, new ->
        updateLifeBalls()
    }

    var oColor : Color? = null;

    override fun tick() {
        super.tick()

        val pos = Vector3(centerX, (y + (height / 2f)), 0f)

        body.setTransform(pos.x, pos.y, Math.toRadians(rotation.toDouble()).toFloat())

        val movingDirection = Direction.fromVector(velocity)

        if (movingDirection != Direction.NONE) {
            if (currentAnimation == null || currentAnimation.direction != movingDirection) {
                getAnimation(AnimationType.RUNNING, movingDirection).reset().play()
            }
        } else if (currentAnimation == null || currentAnimation.type != AnimationType.IDLE) {
            getAnimation(AnimationType.IDLE, lastDirection).reset().play()
        }

        lastDirection = movingDirection
    }

    var dead : Boolean by Delegates.observable(false) {
        d, old, new ->
        if (!old && new) {
            oColor = color;
            color = Color(234 / 255f, 234 / 255f, 32 / 255f, 1f)
        } else if (old && !new) {
            color = oColor;
        }
    }

    private var lifeBall: Array<Entity?> = arrayOfNulls(if (lives < Constants.MAX_LIVES) Constants.MAX_LIVES else lives.toInt())

    override fun onLoad() {
        super.onLoad()

        setScale(0.75f)

        updateLifeBalls()

        val playerDef = BodyDef()

        val pos = Vector3(centerX, (y + (height / 2f)), 0f)

        body = Ghost.getInstance().world.createBody(playerDef)

        val wallBox = CircleShape()
        wallBox.radius = width / 2f

        body.createFixture(wallBox, 0.0f)

        wallBox.dispose()

        body.setTransform(pos.x, pos.y, Math.toRadians(rotation.toDouble()).toFloat())
    }

    fun updateLifeBalls() {
        lifeBall.forEach {
            if (it != null) {
                deattach(it)
                parentScene.removeEntity(it)
            }
        }

        lifeBall = arrayOfNulls<Entity>(if (lives < Constants.MAX_LIVES) Constants.MAX_LIVES else lives.toInt())

        for (i in 0..lives-1) {
            val temp: Entity = Entity.fromImage("sprites/ball.png")

            var newX = centerX - ((width / 1.5f) / 2f)
            newX += (((width / 1.5f) / (Constants.MAX_LIVES - 1)) * i)

            temp.setScale(0.2f)
            temp.setCenter(newX, centerY - 40f)
            temp.color = Color(20 / 255f, 183 / 255f, 52 / 255f, 1f)
            temp.setAlpha(color.a)

            parentScene.addEntity(temp)

            attach(temp)

            lifeBall[i] = temp
        }
    }
}